gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+            Vector x0 = random_vec(m_n_);
+            Vector x00 = random_vec(m_n_);
+            tmp_ritz_vectors.col(0) = perform_op(x0, x00, betas[k]);
+            tmp_ritz_vectors.col(1) =
+                perform_op(tmp_ritz_vectors.col(0), x0, betas[k]);
+          }
+          else {
+            tmp_ritz_vectors.col(j) =
+                perform_op(tmp_ritz_vectors.col(j - 1),
+                           tmp_ritz_vectors.col(j - 2), betas[k]);
+          }
+          const Vector x = tmp_ritz_vectors.col(j);
+          const double up = x.transpose() * A_ * x;
+          const double down = x.transpose().dot(x);
+          const double mu = up / down;
+          if (mu * mu / 4 > maxBeta) {
+            maxIndex = k;
+            maxBeta = mu * mu / 4;
+            break;
+          }
+        }
+      }
+    }
+
+    beta_ = betas[maxIndex];
+  }
+
+  void perturb(int i) {
+    // generate a 0.03*||x_0||_2 as stated in David's paper
+    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::mt19937 generator (seed);
+    std::uniform_real_distribution<double> uniform01(0.0, 1.0);
+
+    int n = m_n_;
+    Vector disturb;
+    disturb.resize(n);
+    disturb.setZero();
+    for (int i =0; i<n; ++i) {
+      disturb(i) = uniform01(generator);
+    }
+    disturb.normalize();
+
+    Vector x0 = ritz_vectors_.col(i);
+    double magnitude = x0.norm();
+    ritz_vectors_.col(i) = x0 + 0.03 * magnitude * disturb;
+  }
+
+  void init(const Vector x0, const Vector x00) {
+    // initialzie ritz eigen values
+    ritz_val_.resize(m_n_);
+    ritz_val_.setZero();
+
+    // initialzie the ritz converged vector
+    ritz_conv_.resize(m_n_);
+    ritz_conv_.setZero();
+
+    // initialzie ritz eigen vectors
+    ritz_vectors_.resize(m_n_, m_n_);
+    ritz_vectors_.setZero();
+    if (accelerated_){
+      ritz_vectors_.col(0) = perform_op(x0, x00);
+      ritz_vectors_.col(1) = perform_op(ritz_vectors_.col(0), x0);
+    } else {
+      ritz_vectors_.col(0) = perform_op(x0);
+      perturb(0);
+    }
+
+    // setting beta
+    if (accelerated_) {
+      Vector init_resid = ritz_vectors_.col(0);
+      const double up = init_resid.transpose() * A_ * init_resid;
+      const double down = init_resid.transpose().dot(init_resid);
+      const double mu =  up/down;
+      beta_ = mu*mu/4;
+      setBeta();
+    }
+    
+  }
+
+  void init() {
+    Vector x0;
+    Vector x00;
+    if (!S_.isZero(0)) {
+      x0 = S_.row(1);
+      x00 = S_.row(0);
+    } else {
+      x0 = random_vec(m_n_);
+      x00 = random_vec(m_n_);
+    }
+    init(x0, x00);
+  }
+
+  bool converged(double tol, int i) {
+    Vector x = ritz_vectors_.col(i);
+    double theta = x.transpose() * A_ * x;
+
+    // store the ritz eigen value
+    ritz_val_(i) = theta;
+
+    // update beta
+    if (accelerated_) beta_ = std::max(beta_, theta * theta / 4);
+
+    Vector diff = A_ * x - theta * x;
+    double error = diff.lpNorm<1>();
```

Comment:
The Ritz residual should be measured using the L2 norm

Reply:
Just to make sure, previous in the email the equation is * || Ax - theta*x || < eps *, so *||* here is l2 norm? If yes, then this should be changed to l2 norm rather than l1.

Reply:
Unless otherwise specified, it's generally understood that ||x|| means L2 norm (standard Euclidean norm).  This should be the L2 norm.